src/reporter/tests/test_notify.py

```diff
@@ -621,4 +621,112 @@ def test_no_value_with_type_for_attributes(service, notification):
     res_get = requests.get(url_new, headers=query_header(service))
     assert res_get.status_code == 200
     assert res_get.json()['values'][0] == None
+    delete_entity_type(service, notification['data'][0]['type'])
+
+
+@pytest.mark.parametrize("service", services)
+def test_issue_382(service, notification):
+    # entity with one Null value and no type
+    notification['data'][0] =     {
+        "id": "urn:ngsi-ld:Test:0002",
+        "type": "Test",
+        "errorNumber": {
+            "type": "Integer",
+            "value": 2
+        },
+        "refVacuumPump": {
+            "type": "Relationship",
+            "value": "urn:ngsi-ld:VacuumPump:FlexEdgePump"
+        },
+        "refOutgoingPallet": {
+            "type": "Array",
+            "value": [
+                "urn:ngsi-ld:Pallet:0003",
+                "urn:ngsi-ld:Pallet:0004"
+            ]
+        }
+    }
+    url = '{}'.format(notify_url)
+    get_url = "{}/entities/urn:ngsi-ld:Test:0002/attrs/errorNumber/value".format(QL_URL)
+    url_new = '{}'.format(get_url)
+    r = requests.post(url, data=json.dumps(notification), headers=notify_header(service))
+    assert r.status_code == 200
+    # Give time for notification to be processed
+    time.sleep(SLEEP_TIME)
+    res_get = requests.get(url_new, headers=query_header(service))
+    assert res_get.status_code == 200
+    assert res_get.json()['values'][0] == 2
+    delete_entity_type(service, notification['data'][0]['type'])
+
+@pytest.mark.parametrize("service", services)
+def test_json_ld(service, notification):
+    # entity with one Null value and no type
+    notification['data'][0] = {
```

Comment:
@jason-fox @kzangeli is ngsi-ld payload representative enough? is there any tricky attribute type we should take into account? is DateTime the only date format that is specified via @value and @type ?

Reply:
Attributes in NGSI-LD (at least right now) can have only three values in their type:
* Property
* Relationship
* GeoProperty

[ Two new attribute types are on their way: "LanguageProperty" and "VocabProperty", but those two are still under discussion in ETSI ISG CIM, so, never mind for now ]

A "Property" attribute has a "value" that can be of any JSON type.
A "Relationship" attribute has an "object" (instead of a "value") that is a string that is a valid URI
A "GeoProperty" has a "value" that is a GeoJSON. Eg: { "type": "Point", coordinates: [] }

There are a few special attributes, like observedAt, observationSpace etc. For that, see the [NGSI-LD spec](https://www.etsi.org/deliver/etsi_gs/CIM/001_099/009/01.03.01_60/gs_CIM009v010301p.pdf).

Reply:
thx @kzangeli

I think current pr would generally cover correctly
* Property
* Relationship
* GeoProperty

The issue is more for the case when the type of Property is not a native JSON Type. So far we implemented support for "DateTime" as by examples, but I could not find other similar not native types, to think of a generalisation. This is important especially if we want to provide backward and forward compatibility (as we are trying to do), where we meant the following:

* backward compatibility: data notified as NGSI-LD should be able to be retrieved as proper NGSI-v2
* forward compatibility: data notified as NGSI-v2 should be able to be retrieved as NGSI-LD

thx for any pointer you can provide to this tricky "custom" data type cases.

Reply:
There's actually a fourth type of attribute/property - TemporalProperty, but that is kind of hidden.
Users aren't allowed to create a property of the type "TemporalProperty", but there is one "timestamp" that can be set by users.

NGSI-LD admits three different Temporal Properties:
* modifiedAt
* crateadAt
* observedAt

The two first are builtins, set by the broker and only observedAt is user-defined.
It's format is key-value basis:

"observedAt": "2020-11-15T10:00:00.000Z"

and, the broker will convert the timestamp into a floating point value.
This to be able to easily compare the timestamp with other timestamps (I imagine QL does the same).
Note though that the timestamps are converted back to ISO8601 form in queries and notifications.

About compound property values with "@type" and "@value", Orion-LD doesn't modify the value but keeps it as is.
This is actually something I've been thinking of taking up with ETSI ISG CIM but I haven't had any time for that still.

For example:
```
{
  "T1": {
    "type": "Property",
    "value": {
      "@type": "DateTime",
      "@value": "2020-..."
    }
}
```
T1 could be saved as a floating point for simpler and faster comparisons.

However, in the notifications that QL receives from the broker, T1 would always be converted back to its original form.
I believe this should be transparent to the subscriber (i.e. QL, Cygnus et al).

About compatibility questions, the easy answer is that:
* if NGSI-LD, simply compact entity types and attr names according to the current @context
* If NGSIv2, send the longnames without compaction.

The "not so easy" answer is that some features (datasetId) are simply not supported by NGSIv2 and there's probably no solution for this (well, if we really want to, there's always a way ...).

An attribute with datasetIds (or: multi-instances) are represented as an array in NGSI-LD, while that would provoke an error in NGSIv2:

"P1": [
  {
    "type": "Property",
    "value": 4,
    "datasetId": "urn:ngsi-ld:xxx:north"
  },
  {
    "type": "Property",
    "value": 9,
    "datasetId": "urn:ngsi-ld:xxx:south"
  }
]

My personal opinion is that this concept of notifying in NGSIv2 is a temporary thing that we may use with some constraints until we implement support for NGSI-LD.

BTW, the whole datasetId concept is quite complex and only partially implemented in Orion-LD as of today.

Right now I can't think of any other problems we may find.

Reply:
thx @kzangeli